src/crimson/os/seastore/transaction_manager.h

```diff
@@ -362,6 +362,73 @@ class TransactionManager : public AsyncCleaner::ExtentCallbackInterface {
     });
   }
 
+  /**
+   * map_existing_extent
+   *
+   * Allocates a new extent at given existing_paddr that must be absolute and
+   * reads disk to fill the extent.
+   * The common usage is that remove the LogicalCachedExtent (laddr~length at paddr)
+   * and map extent to multiple new extents.
+   * placement_hint and generation should follow the original extent.
+   */
+  using map_existing_extent_iertr =
+    alloc_extent_iertr::extend_ertr<Device::read_ertr>;
+  template <typename T>
+  using map_existing_extent_ret =
+    map_existing_extent_iertr::future<TCachedExtentRef<T>>;
+  template <typename T>
+  map_existing_extent_ret<T> map_existing_extent(
+    Transaction &t,
+    laddr_t laddr_hint,
+    paddr_t existing_paddr,
+    extent_len_t length,
+    placement_hint_t placement_hint = placement_hint_t::HOT,
+    reclaim_gen_t gen = DIRTY_GENERATION) {
+    LOG_PREFIX(TransactionManager::map_existing_extent);
+    if (!existing_paddr.is_absolute() ||
+	!t.is_retired(laddr_hint, length, existing_paddr)) {
```

Comment:
Passing the remapped_paddr and remapped_offset is transferring the responsibility of verification the caller, so unless the caller invoke map_existing_extent after invoking dec_ref immediately as test does:

https://github.com/ceph/ceph/blob/db2ec823d8664c20a143dd00107d52b7fe4817c3/src/test/crimson/seastore/test_transaction_manager.cc#L784-L785

the caller should maintain the mapping of remapped_paddr and existing_paddr, see https://github.com/ceph/ceph/pull/45519#discussion_r916691538

Current ObjectDataHandler doesn't hold structure to pass these parameter that it just remove all extents and then alloc new extent, see object_data_handler.cc do_removals and do_insertions.

I can change this if statement to assert to avoid release overhead. It's just a work round, I think it should be a part of lazy read.

Reply:
Not sure if I get your point, I'll provide the commit later.

Reply:
In a word, “mutate-then-split”, “verification of map_existing_extent's parameters” and “lazy read” all need to prove that the split extents in flight are derived from a valid on-disk extent.

To pass « remapped_paddr » and « remapped_offset » to map_existing_extent, we need to keep those two params somewhere, which would probably end up being something the « write_context » or « split_set » that I introduced at an early stage: https://github.com/ceph/ceph/pull/45519#issuecomment-1089948644. However, those structures is hard to be a generic interface, so it was dropped from the PR

Reply:
Yeah, agree that passing remapped_paddr/offset is not a simpler way. I've tried and realized that the split can also happen to the EXIST_CLEAN extents. And that extent will not appear in the transaction's retired_set.